pkg/backend/httpstate/diffs.go

```diff
@@ -31,6 +32,15 @@ import (
 	opentracing "github.com/opentracing/opentracing-go"
 )
 
+func asString(raw json.RawMessage) string {
+	// NOTE: this depends on the fact that the layout of a byte slice and a string share a common header. Both of these
+	// types are laid out as (ptr, len). Once the CLI targets Go 1.20, we should change this to
```

Comment:
> I think we're already publishing binaries with Go 1.20:

We are, but our `go.mod` still specified 1.18, so we can't use 1.20 features by default. I'll try the separate files approach and see if that works.

Reply:
The 1.18 in go.mod doesn't prevent use of new library features. It mostly prevents use of new language features.
e.g., you can't use generics without putting 1.18 in there. (And when the for loop semantics fix lands, that'll be module-scoped based on the version of Go specified in that module's go.mod.)

Reply:
~~The 1.18 in go.mod doesn't prevent use of new library features. It mostly prevents use of new language features.~~

Turns out that `"unsafe"` is treated more as a language feature, not a library function, so you can't use it without go 1.20 in your go.mod same way you can't use generics without 1.18 or higher in your go.mod.

Reply:
So it should be safe to bump the go.mod to 1.20 as long as we're still testing with 1.19.
The `go` directive in the go.mod doesn't actually enforce which version of the compiler is used.
If we use build tags to make sure it still builds with 1.19, and we keep the tests, everything should still work.

But I think we can consider that out of scope for this PR.
Your TODO comment (and possibly an associated issue) should suffice for this PR.
I may follow up on the 1.20 bump later.

Reply:
Hm, also--if the separate files approach would allow us to stick the code that uses the newer library into its own file, then we could avoid the need for a breaking change to the `go.mod` here. I'll look into that as well.

Reply:
Ack. Feel free to try, but I don't think my suggested approach will work after all.
unsafe APIs appear to be treated as a language feature, not like the rest of the standard library where the build tag approach would suffice.

Also, just to clarify: it's not a breaking change to bump the `go` directive in the go.mod. It's not a minimum required version enforcement like other languages. It's just a "code was written assuming these semantics" directive. If a module says `go 1.20` a 1.19 compiler can still try to compile and use it. So if you bump the directive to 1.20, but add a file tagged `//go:build !go1.20` to make the code compatible with 1.19, it'll still work as expected.
All that said, it's probably not worth the effort to do all this just to bump the go directive right now. We're 3 months to the next release, at which point Go 1.20 will be the minimum supported version anyway.

Reply:
Tried the separate files approach out. We could take that approach without modifying our `go.mod` if we updated the build to pass `-gcflags="-lang=go1.20"`. Without that flag, here's what we get:

```sh
httpstate ❯ go build .    
# github.com/pulumi/pulumi/pkg/v3/backend/httpstate
./diffs_post_go120.go:15:23: unsafe.SliceData requires go1.20 or later (-lang was set to go1.18; check go.mod)
```

Reply:
Annoyingly enough, if I remove `diffs_post_go120.go` and leave the pre-120 code in its own file with a build directive, the build still fails when using a 1.20+ toolchain:

```shell
httpstate ❯ go build .                  
# github.com/pulumi/pulumi/pkg/v3/backend/httpstate
./diffs.go:116:43: undefined: unsafeAsString
```

Reply:
I wouldn't use gcflags for that. Anyone consuming this package as a library won't be passing that and will be expecting `go build` to just work, which will base its behavior off go.mod.

Reply:
> Annoyingly enough, if I remove diffs_post_go120.go and leave the pre-120 code in its own file with a build directive, the build still fails when using a 1.20+ toolchain:

Yeah, because we still need the function.
The pre_go120 version still compiles and runs with Go 1.20 just fine, though, right?
We can probably just use that (no build directive) and leave the TODO for the new version once we're minimum Go 1.20.
Basically, just going back to what your original code was here.

Reply:
> Yeah, because we still need the function.

Right. I was hoping that the `//go:build` directive would base its behavior on the Go version stated in `go.mod` (and therefore passed as `-lang` to the compiler), but it doesn't. Seems like a 1.20 toolchain will always define `go1.20` even when it's targeting earlier versions.

On the upside: I have a better understanding of the `go` directive in `go.mod` now. I had assumed that the Go version stated in the `go.mod` was a minimum version requirement to use the module, but that is not the case. It's only used to state the default language version for the module. Older toolchains can still consume and build the module's sources.

IMO https://github.com/golang/go/issues/57001 and https://github.com/golang/go/issues/59033 will improve this situation in that they'll make things behave like I expected.

So all up, I think that what we should do here is:
- For now, leave the code in these changes with the TODO
- For the next set of changes, change the Go version in `go.mod` to 1.20 and add two implementations of `unsafeAsString` and the core diff algorithm: one that targets the old diff package for pre-1.20 versions and one that targets the new diff package for 1.20+.

How does that sound?

Reply:
> Also, just to clarify: it's not a breaking change to bump the go directive in the go.mod. It's not a minimum required version enforcement like other languages. It's just a "code was written assuming these semantics" directive. If a module says go 1.20 a 1.19 compiler can still try to compile and use it. So if you bump the directive to 1.20, but add a file tagged //go:build !go1.20 to make the code compatible with 1.19, it'll still work as expected.

totally missed this comment 🤦

Reply:
Yep! Leave the TODO, bump to 1.20 in a separate change if we want.